### PR TITLE
Edge with UIA: allow activating  buttons in browse mode by walking ancestors until an action succeeds

### LINKED ISSUE #14612: NVDA fails to interact with some web controls on Google pages if using UIA for Microsoft Edge
### Steps to reproduce:
1. From NVDA's advanced settings, set the option "Use UIA with Microsoft Edge and other Chromium based browsers when available" to "Yes".
2. Open Microsoft Edge.
3. Go to youtube.com.
4. Go to the toggle button "Guide" and press ENTER to activate it.

### Actual behavior:
Nothing happens.

### Expected behavior:
The guide panel on the left should appear/disappear, depending on its status and NVDA should report "checked"/"Not checked" as a status for the focused "Guide" toggle button.

### NVDA logs, crash dumps and other attachments:
[NVDA_MSEdge_UIA.log](https://github.com/nvaccess/nvda/files/10664718/NVDA_MSEdge_UIA.log)

#### NVDA installed/portable/running from source:
Installed.

#### NVDA version:
NVDA version: alpha-27627,79ad4f07.

#### Windows version:
Operating system: Windows 11 Pro version 22H2 (build 22621.1105), 64-bit, in Bulgarian with all locale settings set to "Bulgarian".

#### Name and version of other software in use when reproducing the issue:
Microsoft Edge version: 109.0.1518.78, in bulgarian.

#### Does the issue still occur after restarting your computer?
Yes, it does.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes - NVDA 2022.4 and the result is the same.

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes, it is.

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
I haven't done that recently, but previously I have for some other diagnostics and the problem was occurring before that and after it, so this is not relevant.

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #14612

### Summary of the issue:
when accessing Microsoft Edge with UI Automation, NVDA is unable to activate most buttons and controls in browse mode.
This is because NvDA only tries to invoke/select/toggle the deepest element at the browse mode caret position, which may be a text leaf, or span within the interactive control, and not the control itself.

### Description of user facing changes
Microsoft Edge via UI Automation: NVDA can now activate buttons and other controls in browse mode.

### Description of development approach
* UIA NVDAObject's doAction method: if the element cannot be invoked or selected or toggled, then NotImplementedError is raised.
* UIA BrowseMode document's _activateNVDAObject method: Rather than just trying to activate the given object with its doAction method, instead start at the given object and move up its ancestors, until a call to doAction succeeds. I.e. activates the given object or its closest ancestor that supports performing actions.
 
### Testing strategy:
* Ensure NVDA's use UI Automation to access Edge / chromium is set to yes.
* Open Microsoft Edge and navigate to https://www.youtube.com/
* Tab to the Guide button and press enter, and ensure that it activates.

### Known issues with pull request:
None known.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

**Comment by jcsteh** [comment]
Unless I'm missing something, this is not going to work well for non-semantic clickable controls. Obviously, those are authoring error, but there are plenty of them. This might also result in a click potentially getting routed to a far away, irrelevant ancestor that happens to have a click handler for event listening purposes.

With gecko_ia2, when walking ancestors, if doAction fails, NVDA simulates a mouse click if the ancestor has a relevant location and considers that a success. This does occasionally fail and click the wrong place, since location might not be perfect due to bugs, pop-ups obscuring things, etc. That's why Chrome and Firefox implemented click ancestor, but I guess we can't access that with UIA, so click simulation is probably the best we can do.

**Comment by michaelDCurran** [comment]
@jcsteh  I've now moved that gecko_ia2 logic into the base browse mode _activateNVDAObject. I can't see any reason why all implementations can't use that then if we are happy with it.

File: source/NVDAObjects/UIA/__init__.py [hunk 1]
```diff
@@ -2019,11 +2019,13 @@ def doAction(self,index=None):
 		if index==0:
 			if self.UIAInvokePattern:
 				self.UIAInvokePattern.Invoke()
+				return
 			elif self.UIATogglePattern:
 				self.UIATogglePattern.toggle()
+				return
 			elif self.UIASelectionItemPattern:
 				self.UIASelectionItemPattern.select()
-			return
```
Comment (jcsteh) [inline source/NVDAObjects/UIA/__init__.py:2026]: Why this change? I don't think there's a functional difference, but having elif + return is redundant. If you want to keep the early returns in each branch, I'd change the elifs to ifs.

**Review by jcsteh — COMMENTED** [review]
"Happy with it" is probably a stretch. As I noted, it can result in clicking the wrong thing in some cases. However, I think this is probably the most pragmatic solution given the prevalence of non-semantic clickables with ancestor click handlers.

Could this be a problem for non-web contexts; Microsoft Word, etc.?

File: source/NVDAObjects/UIA/__init__.py (hunk 1 above)
Comment (michaelDCurran) [reply source/NVDAObjects/UIA/__init__.py:2026]: Previously it would raise NotImplementedError only if actionIndex was not 0. Now it also raises NotImplementedError if neither invoking, toggling or selecting is possible. At least, that was what I was trying to go for...

**Comment by michaelDCurran** [comment]
@jcsteh  Some quick tests with MS Word seems to suggest things are working okay. It is possibly now more accurate when jumping to references.

File: source/NVDAObjects/UIA/__init__.py (hunk 1 above)
Comment (jcsteh) [reply source/NVDAObjects/UIA/__init__.py:2026]: Ah gotcha. Sorry, missed that subtlety. In that case, please change the elifs to ifs and we're good here.

**Review by jcsteh — APPROVED** [review]

**Review by seanbudd — APPROVED** [review]
Please fix the failing lint check before merging
